### DISCUSSION TITLE
Deleting session in db does not log out the user / invalidate the session.

### DISCUSSION THREAD

**Discussion by jsbrain** [directus/directus#15897] [Archive]
### Describe the Bug

Deleting a user session from the db (manually) still allows subsequent requests via App and local auth.

Deleting a session should immediately block any request to hidden resources, otherwise sessions could be hijacked indefinitely.

After deleting a session from the db and refreshing the App page, the user get's logged out. But continuing using the App without reloading the page let's you do basically anything without restrictions. It seems, the invalidation of the session actually happens in the App itself and is hence not enforced on server side.

If an attacker would hijack the admin session, they could just create another admin user, even if the current session is deleted from the db.

Am I missing something here? Why is this possible?

### To Reproduce

See above.

### What version of Directus are you using?

v9.18.1

### What version of Node.js are you using?

16

### What database are you using?

Postgres 14

### What browser are you using?

Chrome

### How are you deploying Directus?

Docker

**Comment by br41nslug** [comment]
This is a consequence of stateless JWT tokens used for authentication which essentially cannot be revoked until they expire. 
https://fusionauth.io/learn/expert-advice/tokens/revoking-jwts

Edit: The current sessions table does not look like it's used during the authentication process while having all the fields that would be necessary to implement an invalidation feature 🤔

**Reply by jsbrain** [reply]
Probably the "weirdest" thing for me is that deleting the session, which is actually the refresh_token does not render it useless. I mean, why save it when it's not used in the authentication process?

Also, the role is not connected to the session or at least doesn't get re-validated in any way. Which is probably a subsequent issue of the one just described.

~~So when changing the role of a user while they are logged in won't change anything about their permissions. If I set a user to `admin` and change it to `author` the user stays an admin if they never log out. **Using the refresh_token to get a new pairs of token does not update the user to the current role.**~~

~~If there would be at least a check while validating the jwt if the users role is still valid or something like that, that would already be much better.~~

~~What I did in a previous project was simply adding the role and a short id of the actual session id in the jwt payload.~~

~~With this one can simply check if the session or role is still valid and if not invalidate the `refresh_token`. No need to keep track of jwt's or anything, the payload is secure and is being handled on every request anyways.~~

**UPDATE:** I need to confess, there was something wrong with my testing. The role stuff obviously works.

**Reply by jsbrain** [reply]
So to maybe clarify this a little more:

- user logs in
- user is received from db, with current role
- session is being created and written to the db
- access_token is being signed with current roleId (rid) and sessionId (sid) in jwt payload
- refresh_token is being signed with sessionId in jwt payload
- user receives access_token and refresh_token

Now, when a request comes in:

- access_token is verified
- user and session are pulled from db
- check if sessionId matches jwt.payload.sid
- check if roleId matches jwt.payload.rid
- continue or fail (invalidate session)

If a user tries to use the refresh_token after the access_token has been rejected:

- sign the token
- check db: select from session where id = jwt.payload.sid
- return new tokens or invalidate

With this, changing the role of a user would not require any additional work. The user is connected to the session and the role is connected to the user. If any of these change, the validation of the access_token will fail. The next request would be a refresh request, which would sign a new refresh_token and a new access_token including the new role so the permissions of the current user session would change immediately.

Invalidating the refresh_token, hence deleting the session would immediately force the user to log in again.

This would also allow, to decouple the refresh_token from the sessionId, allowing to use other kinds of ids for the session. Not sure if this is a benefit by any means but it's no downside either.

**Reply by jsbrain** [reply]
I only looked briefly into the extensions sdk and afaik I could just implement this there but I thing this should be a core feature and would drastically improve security of Directus.

**Reply by zakwear** [reply]
@jsbrain Checking session role against the user role would break our use of sessions. 

The current behavior: When the user applies a refresh_token, they will get a JWT with the role they are assigned. That's good, as you can default their role to a low level, like Authenticated.

The current setup also permits programmatically creating new Sessions with any arbitrary role, which allows a client to hold a key chain with N-roles. No changing of actual role for the user required. If any of the refresh tokens in those Sessions are used, they'll get a JWT with the Authenticated role (or whatever their assigned/default role is.) Thus, they must ask a Directus Flow or Extension to create a new Session for any non-Authenticated role before expiration. This prevents runaway refresh behavior.
